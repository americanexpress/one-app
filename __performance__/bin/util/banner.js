/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const table = require('markdown-table');
const { Table } = require('console-table-printer');
const {
  testType, time, timeDiff, colors,
} = require('./format');
const { sum } = require('./math');

const metrics = {
  ...require('../k6/metrics'),
  ...require('../prometheus/metrics'),
};

function labelledMetric(metric, data, processor) {
  const { bold } = colors();
  return `${metrics[metric].label}: ${bold(metrics[metric].format(processor(data[metric])))}`;
}

function banner(colData, markdown = false) {
  const rows = [
    colData.map(({ meta }) => testType(meta.type)),
    colData.map(({ meta }) => time(meta.timestamp, 'Completed:')),
    colData.map(({ meta }) => timeDiff(meta.start, meta.end, 'Duration')),
    colData.map(({ data }) => labelledMetric('http_reqs', data, sum)),
    colData.map(({ data }) => labelledMetric('data_received', data, sum)),
    colData.map(({ data }) => labelledMetric('http_req_failed', data, sum)),
  ];

  if (colData.some(({ id }) => !!id)) rows.push(colData.map(({ id }) => id));

  if (markdown) {
    return table([
      colData.map(({ label }) => label),
      ...rows,
    ]);
  }

  const consoleTable = new Table({
    columns: colData.map(({ label }) => ({ name: label, alignment: 'left', color: 'none' })),
    colorMap: { none: '' },
  });

  consoleTable.addRows(rows.map((row) => row.reduce((acc, curr, i) => ({
    ...acc, [colData[i].label]: curr,
  }), {})));

  return consoleTable.render();
}

module.exports = banner;
