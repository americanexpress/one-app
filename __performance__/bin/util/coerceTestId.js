/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const fs = require('node:fs');
const path = require('node:path');
const { diffArrays } = require('diff');
const {
  green, red, dim, underline, strikethrough,
} = require('colorette');
const getResultDir = require('./getResultDir');

function createCoerceTestId(optionName) {
  return function coerceTestId(arg) {
    const dir = /^[./~]/.test(arg) ? path.resolve(process.cwd(), arg) : getResultDir(arg);
    const id = path.basename(dir);

    const dirContents = fs.readdirSync(dir).sort();
    const expectedContents = ['k6-out.json', 'k6-summary.json', 'metrics-k6.json', 'metrics-prom.json'].sort();

    if (!expectedContents.every((file) => dirContents.includes(file))) {
      const contentsDiff = diffArrays(expectedContents, dirContents, { ignoreWhitespace: true });
      const printDiff = contentsDiff.map((part) => (part.added ? strikethrough(green(part.value))
        // eslint-disable-next-line unicorn/no-nested-ternary -- conflicting rules
        : part.removed ? red(part.value) : dim(part.value)));

      throw new Error(`${red(`Bad option ${underline(optionName)}:`)}\n  Invalid test result directory: ${dir}\n\n  ${printDiff.join(', ')}`);
    }

    return { dir, id };
  };
}

module.exports = createCoerceTestId;
