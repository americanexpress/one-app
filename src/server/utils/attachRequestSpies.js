/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import https from 'node:https';
import http from 'node:http';
import url from 'node:url';
import onFinished from 'on-finished';

export function attachSpy(obj, methodName, spy) {
  const origMethod = obj[methodName];

  if (typeof origMethod !== 'function') {
    throw new TypeError(`${methodName} is not a function`);
  }

  if (typeof spy !== 'function') {
    throw new TypeError(`spy must be a function (was "${typeof spy}")`);
  }

  // we're monkeypatching, we need to reassign a property of the obj argument
  // eslint-disable-next-line no-param-reassign -- see above comment
  obj[methodName] = function spiedMethod(...args) {
    let returnValue;
    let originalCalled = false;
    const callOriginal = () => {
      originalCalled = true;
      returnValue = origMethod.apply(obj, args);
      return returnValue;
    };

    spy([...args], callOriginal);

    if (!originalCalled) {
      callOriginal();
    }

    return returnValue;
  };
}

function buildUrlObject(options, defaultProtocol) {
  // TODO: url.parse is deprecated, use new URL() instead
  // this is not a 1:1 replacement and will require changes.
  // Currently the parsed url is used to provide href to
  // loggers outgoingRequestEndSpy function.
  // https://github.com/americanexpress/one-app/blob/main/src/server/utils/logging/setup.js#L42
  // https://nodejs.org/api/url.html#urlparseurlstring-parsequerystring-slashesdenotehost
  const parsedPath = url.parse(options.path || '/');
  const protocol = options.protocol || `${defaultProtocol}:`;
  const urlObject = {
    auth: options.auth,
    hostname: options.hostname || options.host || 'localhost',
    protocol,
    port: options.port || (protocol === 'http:' ? 80 : 443),
    hash: parsedPath.hash,
    path: parsedPath.path,
    pathname: parsedPath.pathname,
    query: parsedPath.query,
    search: parsedPath.search,
  };
  if (
    (protocol === 'http:' && urlObject.port === 80)
    || (protocol === 'https:' && urlObject.port === 443)
  ) {
    urlObject.port = undefined;
  }
  return urlObject;
}

function parseUrl(options, defaultProtocol) {
  // some data is not stored in the clientRequest, have to duplicate some parsing
  // adapted from https://github.com/nodejs/node/blob/894203dae39c7f1f36fc6ba126bb5d782d79b744/lib/_http_client.js#L22
  if (typeof options === 'string') {
    return url.parse(options);
  }

  return url.parse(url.format(buildUrlObject(options, defaultProtocol)));
}

function httpSpy(defaultProtocol, requestSpy, socketCloseSpy) {
  return (args, callOriginal) => {
    const options = args[0];
    const clientRequest = callOriginal();
    const parsedUrl = parseUrl(options, defaultProtocol);

    requestSpy(clientRequest, parsedUrl);

    if (socketCloseSpy) {
      onFinished(clientRequest, () => socketCloseSpy(clientRequest, parsedUrl));
    }
  };
}

export default function attachRequestSpies(requestSpy, socketCloseSpy) {
  if (typeof requestSpy !== 'function') {
    throw new TypeError(`requestSpy must be a function (was "${typeof requestSpy}")`);
  }

  if (socketCloseSpy && (typeof socketCloseSpy !== 'function')) {
    throw new TypeError(
      `socketCloseSpy must be function if provided (was "${typeof socketCloseSpy}")`);
  }

  attachSpy(https, 'request', httpSpy('https', requestSpy, socketCloseSpy));
  attachSpy(http, 'request', httpSpy('http', requestSpy, socketCloseSpy));
}
