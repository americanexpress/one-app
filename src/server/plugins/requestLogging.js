/*
 * Copyright 2022 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import url from 'node:url';
import fp from 'fastify-plugin';
import attachRequestSpies from '../utils/attachRequestSpies';

/*
TIMERS
`$RequestFullDuration` measures how long it took the whole request to be resolved
`$RequestOverhead` measures how long it took from when the request was made
                   until the route handler was called
`$RouteHandler` measures how long the route handler took to provide an output
`$ResponseBuilder` measures how long it took for fastify to build the response
                   after the payload/output from route handled was provided
Why symbols?
It prevents conflicts with existing keys, native or from other plugins
*/
export const $ExternalRequestDuration = Symbol('$ExternalRequestDuration');
export const $RequestFullDuration = Symbol('$RequestFullDuration');
export const $RequestOverhead = Symbol('$RequestOverhead');
export const $RouteHandler = Symbol('$RouteHandler');
export const $ResponseBuilder = Symbol('$ResponseBuilder');

const passThrough = ({ log }) => log;

const UTILS = {
  configureRequestLog: passThrough,
};

function formatProtocol(parsedUrl) {
  const { protocol } = parsedUrl;
  return protocol.replace(/:$/, '');
}

const getLocale = (req) => {
  // TODO: Verify if `store` is available
  if (req.store) {
    const state = req.store.getState();
    return state.getIn(['intl', 'activeLocale']);
  }

  return undefined;
};

const hrtimeToMs = (value) => (value[0] * 1e3) + (value[1] * 1e-6);

const startTimer = (obj, symbol) => {
  const time = process.hrtime();

  // eslint-disable-next-line no-param-reassign
  obj[symbol] = time;
};

const endTimer = (obj, symbol) => {
  const result = process.hrtime(obj[symbol]);
  const ms = hrtimeToMs(result);

  // eslint-disable-next-line no-param-reassign
  obj[symbol] = ms;

  return ms;
};

const buildMetaData = (request, reply) => {
  const { headers } = request;

  return {
    method: request.method,
    correlationId: headers['correlation-id'],
    // null to explicitly signal no value, undefined if not expected for every request
    host: headers.host || null,
    referrer: headers.referer || headers.referrer || null,
    userAgent: headers['user-agent'] || null,
    location: reply.getHeader('location') || undefined,
    forwarded: headers.forwarded || null,
    forwardedFor: headers['x-forwarded-for'] || null,
    locale: getLocale(request),
  };
};

const logClientRequest = (request, reply) => {
  const getTimer = (symbol) => request[symbol];
  const ttfb = getTimer($RequestFullDuration) - getTimer($ResponseBuilder);
  const log = {
    request: {
      direction: 'in',
      protocol: request.protocol,
      address: {
        uri: url.format({
          protocol: request.protocol,
          port: request.port,
          hostname: request.hostname,
          pathname: request.originalUrl,
        }),
      },
      metaData: buildMetaData(request, reply),
      timings: { // https://www.w3.org/TR/navigation-timing/
        duration: Math.round(getTimer($RequestFullDuration)),
        ttfb: ttfb >= 0 ? Math.round(ttfb) : null,
        requestOverhead: Math.round(getTimer($RequestOverhead)),
        routeHandler: Math.round(getTimer($RouteHandler)),
        responseBuilder: Math.round(getTimer($ResponseBuilder)),
      },
      statusCode: reply.statusCode,
      statusText: reply.raw.statusMessage || undefined,
    },
  };

  const configuredLog = UTILS.configureRequestLog({
    req: request,
    res: reply,
    log,
  });

  request.log.info(configuredLog);
};

export const setConfigureRequestLog = (newConfigureRequestLog = passThrough) => {
  UTILS.configureRequestLog = newConfigureRequestLog;
};

/**
 * Fastify Plugin to log requests
 * @param {import('fastify').FastifyInstance} fastify Fastify instance
 * @param {import('fastify').FastifyPluginOptions} opts plugin options
 */
const requestLogging = async (fastify, opts = {}) => {
  // decorators should be initialized per Fastify documentation
  // see https://www.fastify.io/docs/latest/Reference/Decorators/#decoratereplyname-value-dependencies
  fastify.decorateRequest($RequestOverhead, null);
  fastify.decorateRequest($RequestFullDuration, null);
  fastify.decorateRequest($RouteHandler, null);
  fastify.decorateRequest($ResponseBuilder, null);

  if (opts.spy === true) {
    fastify.addHook('onReady', async () => {
      function outgoingRequestSpy(externalRequest) {
        startTimer(externalRequest, $ExternalRequestDuration);
      }

      function outgoingRequestEndSpy(externalRequest, parsedUrl) {
        const { res } = externalRequest;
        const duration = endTimer(externalRequest, $ExternalRequestDuration);
        fastify.log.info({
          request: {
            direction: 'out',
            protocol: formatProtocol(parsedUrl),
            address: {
              uri: parsedUrl.href,
            },
            metaData: {
              method: externalRequest.method,
              // null to explicitly signal no value, undefined if not expected for every request
              correlationId: externalRequest.getHeader('correlation-id') || undefined,
            },
            timings: { duration },
            statusCode: (res && res.statusCode) || null,
            statusText: (res && res.statusMessage) || null,
          },
        });
      }
      attachRequestSpies(outgoingRequestSpy, outgoingRequestEndSpy);
    });
  }

  fastify.addHook('onRequest', async (request) => {
    startTimer(request, $RequestOverhead);
    startTimer(request, $RequestFullDuration);
  });

  fastify.addHook('preHandler', async (request) => {
    endTimer(request, $RequestOverhead);
    startTimer(request, $RouteHandler);
  });

  fastify.addHook('onSend', async (request, _reply, payload) => {
    endTimer(request, $RouteHandler);
    startTimer(request, $ResponseBuilder);

    // Note: `onSend` is meant to be used to modify the payload for the response.
    //        we use it to calculate TTFB (time to first byte) since it's called
    //        by fastify right before `writeHead` is called.
    return payload;
  });

  fastify.addHook('onResponse', async (request, reply) => {
    try {
      endTimer(request, $ResponseBuilder);
      // same as 'reply.getResponseTime()' but our approach
      // helps us to make the code cleaner
      endTimer(request, $RequestFullDuration);

      logClientRequest(request, reply);
    } catch (error) {
      request.log.error(error);
      throw error;
    }
  });
};

export default fp(requestLogging, {
  fastify: '4.x',
  name: 'requestLogging',
});
