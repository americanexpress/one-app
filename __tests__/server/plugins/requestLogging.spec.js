/*
 * Copyright 2022 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import url from 'node:url';
import util from 'node:util';
import attachRequestSpies from '../../../src/server/utils/attachRequestSpies';
import requestLogging, {
  $ExternalRequestDuration,
  $RequestFullDuration,
  $RequestOverhead,
  $RouteHandler,
  $ResponseBuilder,
  setConfigureRequestLog,
} from '../../../src/server/plugins/requestLogging';

jest.mock('pino');
jest.mock('../../../src/server/utils/logging/logger');
jest.mock('../../../src/server/utils/attachRequestSpies');

jest.spyOn(console, 'error').mockImplementation(util.format);

describe('requestLogging', () => {
  jest.spyOn(process, 'hrtime').mockReturnValue([0, 1000]);
  let fastify;

  beforeEach(() => {
    jest.clearAllMocks();
    fastify = {
      log: { info: jest.fn() },
      decorateRequest: jest.fn((name, fn) => {
        fastify[name] = fn;
      }),
      addHook: jest.fn((name, fn) => {
        fastify[name] = fn;
      }),
    };
  });

  it('adds the expected hooks when spying is enabled', async () => {
    await requestLogging(fastify, { spy: true });

    expect(fastify.addHook).toHaveBeenCalledTimes(5);
    expect(fastify.addHook).toHaveBeenCalledWith('onReady', expect.any(Function));
    expect(fastify.addHook).toHaveBeenCalledWith('onRequest', expect.any(Function));
    expect(fastify.addHook).toHaveBeenCalledWith('preHandler', expect.any(Function));
    expect(fastify.addHook).toHaveBeenCalledWith('onSend', expect.any(Function));
    expect(fastify.addHook).toHaveBeenCalledWith('onResponse', expect.any(Function));
  });

  it('does not add the onReady hook when spying is not enabled', async () => {
    await requestLogging(fastify);

    expect(fastify.addHook).toHaveBeenCalledTimes(4);
    expect(fastify.addHook).not.toHaveBeenCalledWith('onReady', expect.any(Function));
    expect(fastify.addHook).toHaveBeenCalledWith('onRequest', expect.any(Function));
    expect(fastify.addHook).toHaveBeenCalledWith('preHandler', expect.any(Function));
    expect(fastify.addHook).toHaveBeenCalledWith('onSend', expect.any(Function));
    expect(fastify.addHook).toHaveBeenCalledWith('onResponse', expect.any(Function));
  });

  describe('onReady', () => {
    it('spies on requests', async () => {
      await requestLogging(fastify, { spy: true });
      await fastify.onReady();
      expect(attachRequestSpies).toHaveBeenCalledTimes(1);
    });

    describe('logging outgoing requests', () => {
      function createExternalRequestAndParsedUrl() {
        const externalRequestHeaders = {
          'correlation-id': '123',
          host: 'example.com',
          referer: 'https://example.com/other-place',
          'user-agent': 'Browser/8.0 (compatible; WXYZ 100.0; Doors LQ 81.4; Boat/1.0)',
        };
        const externalRequest = {
          method: 'GET',
          getHeader: jest.fn((name) => externalRequestHeaders[name]),
          res: {
            statusCode: 200,
            statusMessage: 'OK',
          },
        };
        const parsedUrl = url.parse('https://example.com/place');
        return { externalRequest, parsedUrl, externalRequestHeaders };
      }

      it('starts a timer when the request starts', async () => {
        const { externalRequest } = createExternalRequestAndParsedUrl();
        await requestLogging(fastify, { spy: true });
        await fastify.onReady();
        expect(attachRequestSpies).toHaveBeenCalledTimes(1);
        const outgoingRequestSpy = attachRequestSpies.mock.calls[0][0];
        outgoingRequestSpy(externalRequest);
        expect(externalRequest).toEqual(expect.objectContaining({
          [$ExternalRequestDuration]: [0, 1000],
        }));
      });

      it('is level info', async () => {
        const { externalRequest, parsedUrl } = createExternalRequestAndParsedUrl();
        await requestLogging(fastify, { spy: true });
        await fastify.onReady();
        expect(attachRequestSpies).toHaveBeenCalledTimes(1);
        const outgoingRequestEndSpy = attachRequestSpies.mock.calls[0][1];
        outgoingRequestEndSpy(externalRequest, parsedUrl);
        expect(fastify.log.info).toHaveBeenCalledTimes(1);
        expect(fastify.log.info.mock.calls[0]).toMatchSnapshot();
      });

      it('uses the correlation id header if present', async () => {
        const {
          externalRequest,
          externalRequestHeaders,
          parsedUrl,
        } = createExternalRequestAndParsedUrl();
        externalRequestHeaders['correlation-id'] = '1234';
        await requestLogging(fastify, { spy: true });
        await fastify.onReady();
        expect(attachRequestSpies).toHaveBeenCalledTimes(1);
        const outgoingRequestEndSpy = attachRequestSpies.mock.calls[0][1];
        outgoingRequestEndSpy(externalRequest, parsedUrl);
        expect(fastify.log.info).toHaveBeenCalledTimes(1);
        const entry = fastify.log.info.mock.calls[0][0];
        expect(entry.request.metaData).toHaveProperty('correlationId', '1234');
      });

      it('uses undefined for the correlation id header if not present', async () => {
        // TODO: add holocron API to edit the request so we can add tracing/correlationId headers
        const {
          externalRequest,
          externalRequestHeaders,
          parsedUrl,
        } = createExternalRequestAndParsedUrl();
        delete externalRequestHeaders['correlation-id'];
        await requestLogging(fastify, { spy: true });
        await fastify.onReady();
        expect(attachRequestSpies).toHaveBeenCalledTimes(1);
        const outgoingRequestEndSpy = attachRequestSpies.mock.calls[0][1];
        outgoingRequestEndSpy(externalRequest, parsedUrl);
        const entry = fastify.log.info.mock.calls[0][0];
        expect(entry.request.metaData).toHaveProperty('correlationId', undefined);
      });

      it('uses the status code if present', async () => {
        const { externalRequest, parsedUrl } = createExternalRequestAndParsedUrl();
        externalRequest.res.statusCode = 200;
        await requestLogging(fastify, { spy: true });
        await fastify.onReady();
        expect(attachRequestSpies).toHaveBeenCalledTimes(1);
        const outgoingRequestEndSpy = attachRequestSpies.mock.calls[0][1];
        outgoingRequestEndSpy(externalRequest, parsedUrl);
        const entry = fastify.log.info.mock.calls[0][0];
        expect(entry.request).toHaveProperty('statusCode', 200);
      });

      it('uses null for the status code if not present', async () => {
        const { externalRequest, parsedUrl } = createExternalRequestAndParsedUrl();
        delete externalRequest.res.statusCode;
        await requestLogging(fastify, { spy: true });
        await fastify.onReady();
        expect(attachRequestSpies).toHaveBeenCalledTimes(1);
        const outgoingRequestEndSpy = attachRequestSpies.mock.calls[0][1];
        outgoingRequestEndSpy(externalRequest, parsedUrl);
        const entry = fastify.log.info.mock.calls[0][0];
        expect(entry.request).toHaveProperty('statusCode', null);
      });

      it('uses the status text if present', async () => {
        const { externalRequest, parsedUrl } = createExternalRequestAndParsedUrl();
        externalRequest.res.statusMessage = 'OK';
        await requestLogging(fastify, { spy: true });
        await fastify.onReady();
        expect(attachRequestSpies).toHaveBeenCalledTimes(1);
        const outgoingRequestEndSpy = attachRequestSpies.mock.calls[0][1];
        outgoingRequestEndSpy(externalRequest, parsedUrl);
        const entry = fastify.log.info.mock.calls[0][0];
        expect(entry.request).toHaveProperty('statusText', 'OK');
      });

      it('uses null for the status text if not present', async () => {
        const { externalRequest, parsedUrl } = createExternalRequestAndParsedUrl();
        delete externalRequest.res.statusMessage;
        await requestLogging(fastify, { spy: true });
        await fastify.onReady();
        expect(attachRequestSpies).toHaveBeenCalledTimes(1);
        const outgoingRequestEndSpy = attachRequestSpies.mock.calls[0][1];
        outgoingRequestEndSpy(externalRequest, parsedUrl);
        const entry = fastify.log.info.mock.calls[0][0];
        expect(entry.request).toHaveProperty('statusText', null);
      });
    });
  });

  describe('onRequest', () => {
    it('starts Request Overhead and Request Full Duration timers', async () => {
      await requestLogging(fastify);

      expect(process.hrtime).toHaveBeenCalledTimes(0);

      const request = {};

      await fastify.onRequest(request);

      expect(process.hrtime).toHaveBeenCalledTimes(2);
      expect(request).toEqual({
        [$RequestOverhead]: [0, 1000],
        [$RequestFullDuration]: [0, 1000],
      });
    });
  });

  describe('preHandler', () => {
    it('ends Request Overhead timer and starts Route Handler timer', async () => {
      await requestLogging(fastify);

      expect(process.hrtime).toHaveBeenCalledTimes(0);

      const request = {};

      await fastify.preHandler(request);

      expect(process.hrtime).toHaveBeenCalledTimes(2);
      expect(request).toEqual({
        [$RequestOverhead]: 0.001,
        [$RouteHandler]: [0, 1000],
      });
    });
  });

  describe('onSend', () => {
    it('ends Route Handler timer and starts Response Builder timer with unmodified payload', async () => {
      const payload = { intact: true };

      await requestLogging(fastify);

      expect(process.hrtime).toHaveBeenCalledTimes(0);

      const request = {};

      expect(await fastify.onSend(request, null, payload)).toEqual(payload);
      expect(process.hrtime).toHaveBeenCalledTimes(2);
      expect(request).toEqual({
        [$RouteHandler]: 0.001,
        [$ResponseBuilder]: [0, 1000],
      });
    });
  });

  describe('onResponse', () => {
    it('ends Response Builder timer and starts Request Full Duration timer', async () => {
      await requestLogging(fastify);

      expect(process.hrtime).toHaveBeenCalledTimes(0);

      const request = {
        headers: {},
        log: { info: jest.fn() },
      };
      const reply = {
        getHeader: jest.fn(),
        raw: {},
      };

      await fastify.onResponse(request, reply);

      expect(process.hrtime).toHaveBeenCalledTimes(2);
      expect(request).toEqual({
        headers: {},
        log: request.log,
        [$ResponseBuilder]: 0.001,
        [$RequestFullDuration]: 0.001,
      });
    });

    describe('logs the request', () => {
      afterEach(() => {
        setConfigureRequestLog(({ log }) => log);
      });

      it('no headers only fallbacks', async () => {
        await requestLogging(fastify);

        expect(process.hrtime).toHaveBeenCalledTimes(0);

        const request = {
          headers: {},
          log: {
            error: jest.fn(),
            info: jest.fn(),
          },
        };
        const reply = {
          getHeader: jest.fn(),
          raw: {},
        };

        await fastify.onRequest(request, reply);
        await fastify.preHandler(request, reply);
        await fastify.onSend(request, reply);
        await fastify.onResponse(request, reply);

        expect(request.log.info).toHaveBeenCalledTimes(1);
        expect(request.log.info).toHaveBeenCalledWith({
          request: {
            address: {
              uri: '',
            },
            direction: 'in',
            metaData: {
              correlationId: undefined,
              forwarded: null,
              forwardedFor: null,
              host: null,
              locale: undefined,
              location: undefined,
              method: undefined,
              referrer: null,
              userAgent: null,
            },
            protocol: undefined,
            statusCode: undefined,
            statusText: undefined,
            timings: {
              duration: 0,
              requestOverhead: 0,
              responseBuilder: 0,
              routeHandler: 0,
              ttfb: 0,
            },
          },
        });
      });

      it('all headers and request keys present', async () => {
        await requestLogging(fastify);

        expect(process.hrtime).toHaveBeenCalledTimes(0);

        const request = {
          headers: {
            'correlation-id': 123,
            host: 'localhost',
            referer: 'referer',
            'user-agent': 'chrome',
            forwarded: 'forwarded',
            'x-forwarded-for': 'x-forwarded-for',
          },
          protocol: 'https',
          port: 1234,
          hostname: 'amex.com',
          originalUrl: 'amex.com',
          method: 'GET',
          log: { info: jest.fn() },
        };
        const reply = {
          getHeader: jest.fn((header) => {
            switch (header) {
              case 'location': {
                return 'some-location';
              }
              default: {
                return null;
              }
            }
          }),
          raw: {
            statusMessage: 'Custom Message',
          },
          statusCode: 500,
        };

        await fastify.onRequest(request, reply);
        await fastify.preHandler(request, reply);
        await fastify.onSend(request, reply);
        await fastify.onResponse(request, reply);

        expect(request.log.info).toHaveBeenCalledTimes(1);
        expect(request.log.info).toHaveBeenCalledWith({
          request: {
            address: {
              uri: 'https://amex.com:1234/amex.com',
            },
            direction: 'in',
            metaData: {
              correlationId: 123,
              forwarded: 'forwarded',
              forwardedFor: 'x-forwarded-for',
              host: 'localhost',
              locale: undefined,
              location: 'some-location',
              method: 'GET',
              referrer: 'referer',
              userAgent: 'chrome',
            },
            protocol: 'https',
            statusCode: 500,
            statusText: 'Custom Message',
            timings: {
              duration: 0,
              requestOverhead: 0,
              responseBuilder: 0,
              routeHandler: 0,
              ttfb: 0,
            },
          },
        });
      });

      it('activeLocale in intl from store', async () => {
        await requestLogging(fastify);

        expect(process.hrtime).toHaveBeenCalledTimes(0);

        const request = {
          headers: {},
          store: {
            getState: () => ({
              getIn: () => 'us_en',
            }),
          },
          log: { info: jest.fn() },
        };
        const reply = {
          getHeader: jest.fn(),
          raw: {},
        };

        await fastify.onRequest(request, reply);
        await fastify.preHandler(request, reply);
        await fastify.onSend(request, reply);
        await fastify.onResponse(request, reply);

        expect(request.log.info).toHaveBeenCalledTimes(1);
        expect(request.log.info).toHaveBeenCalledWith({
          request: {
            address: {
              uri: '',
            },
            direction: 'in',
            metaData: {
              correlationId: undefined,
              forwarded: null,
              forwardedFor: null,
              host: null,
              locale: 'us_en',
              location: undefined,
              method: undefined,
              referrer: null,
              userAgent: null,
            },
            protocol: undefined,
            statusCode: undefined,
            statusText: undefined,
            timings: {
              duration: 0,
              requestOverhead: 0,
              responseBuilder: 0,
              routeHandler: 0,
              ttfb: 0,
            },
          },
        });
      });

      it('custom configureRequestLog', async () => {
        await requestLogging(fastify);

        const mutateLog = jest.fn(() => 'log changed');

        setConfigureRequestLog(mutateLog);

        const request = {
          headers: {},
          raw: 'raw-request',
          log: { info: jest.fn() },
        };
        const reply = {
          getHeader: jest.fn(),
          raw: 'raw-response',
        };

        await fastify.onRequest(request, reply);
        await fastify.preHandler(request, reply);
        await fastify.onSend(request, reply);
        await fastify.onResponse(request, reply);

        expect(mutateLog).toHaveBeenCalledWith({
          req: request,
          res: reply,
          log: {
            request: {
              address: {
                uri: '',
              },
              direction: 'in',
              metaData: {
                correlationId: undefined,
                forwarded: null,
                forwardedFor: null,
                host: null,
                locale: undefined,
                location: undefined,
                method: undefined,
                referrer: null,
                userAgent: null,
              },
              protocol: undefined,
              statusCode: undefined,
              statusText: undefined,
              timings: {
                duration: 0,
                requestOverhead: 0,
                responseBuilder: 0,
                routeHandler: 0,
                ttfb: 0,
              },
            },
          },
        });
        expect(request.log.info).toHaveBeenCalledTimes(1);
        expect(request.log.info).toHaveBeenCalledWith('log changed');
      });

      it('negative ttfb', async () => {
        await requestLogging(fastify);

        const request = {
          headers: {},
          log: { info: jest.fn() },
        };
        const reply = {
          getHeader: jest.fn(),
          raw: {},
        };

        process.hrtime.mockImplementationOnce(() => [0, 200]);
        process.hrtime.mockImplementationOnce(() => [0, 100]);

        await fastify.onResponse(request, reply);

        expect(request.log.info).toHaveBeenCalledTimes(1);
        expect(request.log.info).toHaveBeenCalledWith({
          request: {
            address: {
              uri: '',
            },
            direction: 'in',
            metaData: {
              correlationId: undefined,
              forwarded: null,
              forwardedFor: null,
              host: null,
              locale: undefined,
              location: undefined,
              method: undefined,
              referrer: null,
              userAgent: null,
            },
            protocol: undefined,
            statusCode: undefined,
            statusText: undefined,
            timings: {
              duration: 0,
              requestOverhead: Number.NaN,
              responseBuilder: 0,
              routeHandler: Number.NaN,
              ttfb: null,
            },
          },
        });
      });
    });

    it('catches and logs errors', async () => {
      const request = {
        headers: {},
        log: { error: jest.fn() },
      };
      const reply = {
        getHeader: jest.fn(),
        raw: {},
      };

      await requestLogging(fastify);
      const boomError = new Error('boom');
      setConfigureRequestLog(() => {
        throw boomError;
      });
      await expect(() => fastify.onResponse(request, reply)).rejects.toEqual(boomError);
      expect(request.log.error).toHaveBeenCalledWith(boomError);
    });
  });

  describe('setConfigureRequestLog', () => {
    it('resets to default', async () => {
      const request = {
        log: { info: jest.fn() },
        headers: {},
      };
      const reply = {
        getHeader: jest.fn(),
        raw: {},
      };

      await requestLogging(fastify);
      setConfigureRequestLog(() => {
        throw new Error('shh');
      });
      setConfigureRequestLog();

      await fastify.onResponse(request, reply);

      expect(request.log.info.mock.calls[0][0]).toMatchInlineSnapshot(`
        {
          "request": {
            "address": {
              "uri": "",
            },
            "direction": "in",
            "metaData": {
              "correlationId": undefined,
              "forwarded": null,
              "forwardedFor": null,
              "host": null,
              "locale": undefined,
              "location": undefined,
              "method": undefined,
              "referrer": null,
              "userAgent": null,
            },
            "protocol": undefined,
            "statusCode": undefined,
            "statusText": undefined,
            "timings": {
              "duration": 0,
              "requestOverhead": NaN,
              "responseBuilder": 0,
              "routeHandler": NaN,
              "ttfb": 0,
            },
          },
        }
      `);
    });
  });
});
