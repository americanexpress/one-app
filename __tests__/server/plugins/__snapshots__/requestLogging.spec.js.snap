// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requestLogging onReady logging outgoing requests is level info 1`] = `
[
  {
    "request": {
      "address": {
        "uri": "https://example.com/place",
      },
      "direction": "out",
      "metaData": {
        "correlationId": "123",
        "method": "GET",
      },
      "protocol": "https",
      "statusCode": 200,
      "statusText": "OK",
      "timings": {
        "duration": 0.001,
      },
    },
  },
]
`;
